crypto/signatures.go

```diff
@@ -25,26 +25,74 @@ var (
 	ErrInvalidSignature = errors.New("invalid signature")
 )
 
-// GenerateKeyPair creates a public-secret keypair that can be used to sign and
-// verify messages.
-func GenerateSignatureKeys() (sk SecretKey, pk PublicKey, err error) {
-	var entropy [EntropySize]byte
-	_, err = rand.Read(entropy[:])
+// KeyPairGenerator allows the caller to generate public-secret key pairs.
+type KeyPairGenerator interface {
+	Generate() (SecretKey, PublicKey, error)
+	GenerateDeterministic(entropy [EntropySize]byte) (SecretKey, PublicKey)
+}
+
+// entropySource allows the caller to retrieve an array of bytes populated to
+// random values.
+type entropySource interface {
+	getEntropy() ([EntropySize]byte, error)
+}
+
+// keyDeriver allows the caller to generate a public-secret key pair based on
+// provided entropy.
+type keyDeriver interface {
+	deriveKeyPair([EntropySize]byte) (SecretKey, PublicKey)
+}
+
+// stdGenerator is an implementation of KeyPairGenerator, allowing the caller
+// to generate public-secret key pairs.
+type stdGenerator struct {
+	es entropySource
+	kd keyDeriver
+}
+
+// Generate creates a public-secret keypair that can be used to sign and verify
+// messages.
+func (sg stdGenerator) Generate() (sk SecretKey, pk PublicKey, err error) {
+	entropy, err := sg.es.getEntropy()
 	if err != nil {
 		return
 	}
-
-	skPointer, pkPointer := ed25519.GenerateKey(entropy)
-	return *skPointer, *pkPointer, nil
+	sk, pk = sg.kd.deriveKeyPair(entropy)
+	return sk, pk, nil
 }
 
-// DeterministicSignatureKeys generates keys deterministically using the input
+// GenerateDeterministic generates keys deterministically using the input
 // entropy. The input entropy must be 32 bytes in length.
-func DeterministicSignatureKeys(entropy [EntropySize]byte) (SecretKey, PublicKey) {
+func (sg stdGenerator) GenerateDeterministic(entropy [EntropySize]byte) (SecretKey, PublicKey) {
+	return sg.kd.deriveKeyPair(entropy)
+}
+
+// randSource is an implementation of entropySource that uses rand.Read to
+// generate random bytes.
+type randSource struct{}
+
+// getEntropy returns an array of bytes with random values.
+func (rs randSource) getEntropy() (entropy [EntropySize]byte, err error) {
+	if _, err := rand.Read(entropy[:]); err != nil {
+		return entropy, err
+	}
+	return entropy, nil
```

Comment:
I understand why we don't want named return values for `deriveKeyPair` but why wouldn't we want them in `getEntropy`? It otherwise forces us to add a `var entropy [EntropySize]byte` line.